plutus-core/untyped-plutus-core/src/Language/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -0,0 +1,510 @@
+import qualified Data.DList                                       as DList
+import           Data.Hashable                                    (Hashable)
+import           Data.STRef
+import           Data.Text.Prettyprint.Doc
+
+{- Note [Scoping]
+   The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
+   prerequisite. The CEK machine correctly handles name shadowing.
+-}
+
+data ExBudgetCategory fun
+    = BConst
+    | BVar
+    | BLamAbs
+    | BApply
+    | BDelay
+    | BForce
+    | BError
+    | BBuiltin         -- Cost of evaluating a Builtin AST node
+    | BBuiltinApp fun  -- Cost of evaluating a fully applied builtin function
+    | BAST
+    deriving stock (Show, Eq, Ord, Generic)
+    deriving anyclass (NFData, Hashable)
+instance Show fun => Pretty (ExBudgetCategory fun) where
+    pretty = viaShow
+
+instance ExBudgetBuiltin fun (ExBudgetCategory fun) where
+    exBudgetBuiltin = BBuiltinApp
+
+type TermWithMem uni fun = Term Name uni fun ExMemory
+
+{- Note [Arities in VBuiltin]
+The VBuiltin value below contains two copies of the arity (list of
+TypeArg/TermArg pairs) for the relevant builtin.  The second of these
+is consumed as the builtin is instantiated and applied to arguments,
+to check that type and term arguments are interleaved correctly.  The
+first copy of the arity is left unaltered and only used by
+dischargeCekValue if we have to convert the frame back into a term
+(see mkBuiltinApplication).  An alternative would be to look up the
+full arity in mkBuiltinApplication, but that would require a lot of
+things to be monadic (including the PrettyBy instance for CekValue,
+which is a problem.)
+-}
+
+-- 'Values' for the modified CEK machine.
+data CekValue uni fun =
+    VCon ExMemory (Some (ValueOf uni))
+  | VDelay ExMemory (TermWithMem uni fun) (CekValEnv uni fun)
+  | VLamAbs ExMemory Name (TermWithMem uni fun) (CekValEnv uni fun)
+  | VBuiltin            -- A partial builtin application, accumulating arguments for eventual full application.
+      ExMemory
+      fun
+      Arity             -- Sorts of arguments to be provided (both types and terms): *don't change this*.
+      Arity             -- A copy of the arity used for checking applications/instantiatons: see Note [Arities in VBuiltin]
+      Int               -- The number of @force@s to apply to the builtin.
+                        -- We need it to construct a term if the machine is returning a stuck partial application.
+      [CekValue uni fun]    -- Arguments we've computed so far.
+    deriving (Show, Eq) -- Eq is just for tests.
+
+type CekValEnv uni fun = UniqueMap TermUnique (CekValue uni fun)
+
+-- | The CEK machine is parameterized over a @spendBudget@ function that has (roughly) the same type
+-- as the one from the 'SpendBudget' class (and so the @SpendBudget@ instance for 'CekCarryingM'
+-- defers to the function stored in the environment). This makes the budgeting machinery extensible
```

Comment:
No, because `Language.PlutusCore.Constant.Apply.applyTypeSchemed` needs to know how to access the spending function. We could provide the spending function as an argument to `applyTypeSchemed` though, but I'd defer this change to a future PR.